ipn/ipnserver/server.go

```diff
@@ -581,6 +583,76 @@ func (s *server) writeToClients(n ipn.Notify) {
 	}
 }
 
+// tryWindowsAppDataMigration attempts to copy the Windows state storage
+// directory from its old location to the new location. (Issue 2856)
+//
+// Tailscale 1.14 and before stored state under %LocalAppData%
+// (usually "C:\WINDOWS\system32\config\systemprofile\AppData\Local"
+// when tailscaled.exe is running as a non-user system service).
+// However it is frequently cleared for almost any reason: Windows
+// updates, System Restore, even various System Cleaner utilities.
+//
+// On failure, return the old location as a fallback where appropriate.
+func tryWindowsAppDataMigration(path string) string {
+	if path != paths.DefaultTailscaledStateFile() {
+		// If they're specifying a non-default path, just trust that they know
+		// what they are doing.
+		return path
+	}
+
+	oldFile := filepath.Join(os.Getenv("LocalAppData"), "Tailscale", "server-state.conf")
+	_, err := os.Stat(oldFile)
+	if os.IsNotExist(err) {
+		// Common case for a freshly installed system.
+		return path
+	}
+	if err != nil {
+		log.Printf("tryWindowsAppDataMigration failed; AppDataLocal: %v", err)
+		return path
+	}
+
+	newFile := filepath.Join(os.Getenv("ProgramData"), "Tailscale", "server-state.conf")
+	_, err = os.Stat(newFile)
+	if err == nil {
+		// Common case for a system which was upgraded from 1.14, we copied the file
+		// to %PROGRAMDATA% while leaving the original in %LOCALAPPDATA%.
+		return path
+	}
+	if !os.IsNotExist(err) {
+		log.Printf("tryWindowsAppDataMigration failed; new file: %v", err)
+		return path
+	}
+
+	// copy server-state.conf from %LOCALAPPDATA% to %PROGRAMDATA%
+	contents, err := os.ReadFile(oldFile)
```

Comment:
no need to stat the oldFile first. You can just jump right to ReadFile it.

Reply:
Done